Fix coreIPC fuzz blocker: NetworkStorageManager::didGenerateIndexKeyForRecord

roberto-apple opened the pull request:
#### ab75ad576ed7e70d59e5a0751752c73cb2310e54
<pre>
Fix coreIPC fuzz blocker: NetworkStorageManager::didGenerateIndexKeyForRecord
<a href="https://bugs.webkit.org/show_bug.cgi?id=294375">https://bugs.webkit.org/show_bug.cgi?id=294375</a>
<a href="https://rdar.apple.com/152492569">rdar://152492569</a>

Reviewed by Sihui Liu.

Check whether key type of WebCore::IDBResourceIdentifier is empty before
attempting to use it to retrieve a value from a HashMap.

* LayoutTests/ipc/networkstoragemanager-didgenerateindexkeyforrecord-notransactionid-expected.txt: Added.
* LayoutTests/ipc/networkstoragemanager-didgenerateindexkeyforrecord-notransactionid.html: Added.
* Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.cpp:
(WebKit::IDBStorageRegistry::transaction):

Canonical link: <a href="https://commits.webkit.org/296198@main">https://commits.webkit.org/296198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f8eb14ab939a4a9dfc4b27a34453ecd4c036db7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112800 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58125 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35768 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81681 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110512 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22157 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96967 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62061 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21594 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15106 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57564 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91527 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15140 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115901 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90712 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93217 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90453 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23089 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35375 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13156 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30414 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34572 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40128 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34318 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35981 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.cpp:
```diff
@@ -100,7 +100,7 @@ WebCore::IDBServer::UniqueIDBDatabaseConnection* IDBStorageRegistry::connection(
 WebCore::IDBServer::UniqueIDBDatabaseTransaction* IDBStorageRegistry::transaction(WebCore::IDBResourceIdentifier identifier)
 {
-    return m_transactions.get(identifier).get();
+    return m_transactions.contains(identifier) ? m_transactions.get(identifier).get() : nullptr;
```
It seems like this exact same thing can happen to `::connection` above.

cc @szewai @youennf

roberto-apple replied:
After taking a look it seems that might neccessarily not be the case. Regarding IDBStorageRegistry::connection(), the declaration of m_connections is
`HashMap<WebCore::IDBDatabaseConnectionIdentifier, WeakPtr<WebCore::IDBServer::UniqueIDBDatabaseConnection>> m_connections;` where the key type is `WebCore::IDBDatabaseConnectionIdentifier`. [Link](https://github.com/WebKit/WebKit/blob/main/Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.h#L63)

To reproduce the same type of crash from the radar for IDBStorageRegistry::connection() (i.e., where the get() method of HashMap is passed a key type with an "empty value" and triggers an [internal release assert](https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/HashTable.h#L371)) an IDBDatabaseConnectionIdentifier would have to be created with a value of 0 and then used as the argument in the get() method. However, IDBDatabaseConnectionIdentifier internally is an `ObjectIdentifierGenericBase<unsigned long long>` and a value of 0 is considered invalid and can't be instantiated (see [here](https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/ObjectIdentifier.h#L86) and [here](https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/ObjectIdentifier.h#L153)), at least with the coreipc.js testing api to my knowledge.

Perhaps someone else might know?

szewai replied:
Yeah I think your analysis makes sense.
To avoid double lookup you might:
```
if (auto transaction = m_transactions.getOptional(identifier))
    return transaction->get();
return nullptr;
```

roberto-apple replied:
Unfortunately it seems that the HashMap::getOptional(key) approach lead to the same kind of release assert as calling HashMap::get(key) with an "empty" key:

IDBStorageRegistry::transaction(key)
->HashMap::getOptional(key)
 -->HashMap::lookup(key)
 --->HashMap::inlineLookup(key)
 ---->HashMap::checkHashTableKey(key)
 ----->// release assert for empty value for key of type WebCore::IDBResourceIdentifier
 
 To avoid the double lookup, looks like simplest approach is to revert to original code and just add an empty check for the key (didn't notice the isEmpty() method before on IDBResourceIdentifier).

roberto-apple committed:
Fix coreIPC fuzz blocker: NetworkStorageManager::didGenerateIndexKeyForRecord
https://bugs.webkit.org/show_bug.cgi?id=294375
rdar://152492569

Reviewed by Sihui Liu.

Check whether key type of WebCore::IDBResourceIdentifier is empty before
attempting to use it to retrieve a value from a HashMap.

* LayoutTests/ipc/networkstoragemanager-didgenerateindexkeyforrecord-notransactionid-expected.txt: Added.
* LayoutTests/ipc/networkstoragemanager-didgenerateindexkeyforrecord-notransactionid.html: Added.
* Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.cpp:
(WebKit::IDBStorageRegistry::transaction):

Canonical link: https://commits.webkit.org/296198@main
